docs/rfc/rfc-103-incoming-txs-when-catching-up.md

```diff
@@ -0,0 +1,177 @@
+- Should nodes or clients send transactions when they suspect the receiving node
+  is not initialized?
+
+## Background
+
+### Life cycle of a node
+
+A CometBFT node transitions through the following stages: set up,
+initialization, catch up, and consensus.
+
+When a node is executed, it first reads its configuration files and *sets up*
+all its components, including reactors, stores, P2P connections, and RPC
+services. Then, the node starts running and turns on all the components. At this
+point we say that the node is *initialized*. The mempool reactor is running and
+the [RPC endpoints](#rpc) are open, allowing the node to [receive](#receive) and
+process transactions from other nodes and clients.
+
+The node could start the consensus protocol immediately to reconstruct the block
+history, but this could take a very long time. BlockSync and StateSync are much
+faster mechanisms to bootstrap the node and reach the latest heights of the
+chain. When a node is performing BlockSync or StateSync, it is in *catch up*
+mode. Once synchronization finishes, the node is ready to fully participate in
+consensus, so it switches to *consensus* mode, which should never leave unless
+the node crashes.
+
+### Transaction validation with CheckTx
+
+The `CheckTx` [method](#check-tx) acts as a gatekeeper to the mempool, which
+only accepts valid transactions. The validity of a transaction is defined
+deterministically by the application, but it may depend on the application
+state, which may change at every height.
+
+`CheckTx` is called:
+- from the [RPC endpoints](#rpc) `broadcast_tx_async`, `broadcast_tx_sync`, and
+  `broadcast_tx_commit`, and
+- when the node [receives](#receive) a `Txs` message from a peer: for each
+  transaction in `Txs` there will be a call to `CheckTx`.
+
+`CheckTx` puts the transaction in the cache, sends a `RequestCheckTx` ABCI
+message to the application, and sets a callback to process a `ResponseCheckTx`.
+If the application deems the transaction as valid, `CheckTx` stores the
+transaction in the mempool and notifies the consensus reactor that there are
+transactions available to be [put in a new block](#reap). Otherwise it does not
+add the transaction to the mempool and removes it from the cache.
+
+### Mempool cache
+
+The mempool implements a cache of received transactions that it uses to discard
+already-seen transactions, thus avoiding processing duplicates.
+
+Transactions remain in the cache until the application replies in
+`ResponseCheckTx` that a transaction is invalid. This can occur the first time a
+transaction is received and processed by `CheckTx`, or when all the transactions
+in the mempool need to be rechecked after updating the mempool. Another instance
+is when the consensus reactor [updates](#update) the mempool after finalizing a
+block: if there was an error while executing a transaction, as indicated by the
+application in `ResponseFinalizeBlock`, then the transaction is removed from the
+cache.
```

Comment:
This part I don't understand.

> Another instance is when the consensus reactor [updates](#update) the mempool after finalizing a block: if there was an error while executing a transaction, as indicated by the application in `ResponseFinalizeBlock`, then the transaction is removed from the cache.

Isn't it the case that a transaction that is part of a finalized block should not be in the cache? Not sure I understand how can it be correct to have the same transaction both in a block being executed and part of the cache.

Reply:
Right, this part is not correct. It should say something like "when the consensus reactor is in the process of commiting a block and needs to update the mempool". Updating the mempool happens right at the end of blockExec.Commit, so technically the block is finalized and committed at this moment.